homeassistant/components/tag/__init__.py

```diff
@@ -90,21 +125,231 @@ async def _update_data(self, item: dict, update_data: dict) -> dict:
+                )
+            data.pop(CONF_NAME)
+
         return data
 
 
+class TagDictStorageCollectionWebsocket(
+    collection.StorageCollectionWebsocket[collection.DictStorageCollection]
+):
+    """Class to expose tag storage collection management over websocket."""
+
+    def __init__(
+        self,
+        storage_collection: collection.DictStorageCollection,
+        api_prefix: str,
+        model_name: str,
+        create_schema: ConfigType,
+        update_schema: ConfigType,
+    ) -> None:
+        """Initialize a websocket for tag."""
+        super().__init__(
+            storage_collection, api_prefix, model_name, create_schema, update_schema
+        )
+        self.entity_registry = er.async_get(storage_collection.hass)
+
+    @callback
+    def ws_list_item(
+        self, hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict
+    ) -> None:
+        """List items specifically for tag.
+
+        Provides name from entity_registry instead of storage collection.
+        """
+        tag_items = self.storage_collection.async_items()
+        for item in tag_items:
+            if (
+                entity_id := self.entity_registry.async_get_entity_id(
+                    DOMAIN, DOMAIN, item[TAG_ID]
+                )
+            ) and (entity := self.entity_registry.async_get(entity_id)):
+                item[CONF_NAME] = entity.name or entity.original_name
+        if _LOGGER.isEnabledFor(logging.DEBUG):
+            _LOGGER.debug("Listing tags %s", tag_items)
+        connection.send_result(msg["id"], tag_items)
+
+    async def ws_create_item(
+        self, hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict
+    ) -> None:
+        """Create a tag item.
+
+        Provides name from entity registry.
+        """
+        try:
+            data = dict(msg)
+            data.pop("id")
+            data.pop("type")
+            item = await self.storage_collection.async_create_item(data)
+            if (
+                entity_id := self.entity_registry.async_get_entity_id(
+                    DOMAIN, DOMAIN, item[TAG_ID]
+                )
+            ) and (entity := self.entity_registry.async_get(entity_id)):
+                item[CONF_NAME] = entity.name or entity.original_name
+            if _LOGGER.isEnabledFor(logging.DEBUG):
+                _LOGGER.debug("Creating tag %s", item)
+            connection.send_result(msg["id"], item)
+        except vol.Invalid as err:
+            connection.send_error(
+                msg["id"],
+                websocket_api.const.ERR_INVALID_FORMAT,
+                humanize_error(data, err),
+            )
+        except ValueError as err:
+            connection.send_error(
+                msg["id"], websocket_api.const.ERR_INVALID_FORMAT, str(err)
+            )
+
+    async def ws_update_item(
+        self, hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict
+    ) -> None:
```

Comment:
Same comment as for ws_create_item; we should not have to override this method?

Reply:
Same as previous